docs/content/sips/010-key-value.md

```diff
@@ -0,0 +1,60 @@
+title = "SIP 010 - Key-Value Storage"
+template = "main"
+date = "2023-01-23T01:01:01Z"
+---
+
+Summary: Provide a generic interface for access to various key-value storage systems.
+
+Owner(s): joel.dice@fermyon.com
+
+Created: Jan 23, 2023
+
+## Background
+
+Spin should have native support for using a variety of key-value (KV) datastores.
+
+Currently, users can use Redis for persistent key-value storage for Spin applications. This [documentation](https://developer.fermyon.com/cloud/data-redis.md) walks through how to use the existing Spin Redis SDK which leverages the `[outbound-redis.wit` interface](https://github.com/fermyon/spin/blob/main/wit/ephemeral/outbound-redis.wit). In contrast to the Redis interface, we are proposing a more general-purpose interface which can be used with a variety of implementations.
+
+## Proposal
+
+In order to support key-value stores, the following need to be added to Spin:
+
+- A `WIT` file that defines the key-value interface
+- SDK implementations for various programming languages
+- A default local key-value store
```

Comment:
We should mention that, as a new import in the SDK, this has interactions with cloud deployment:

1. An application that uses KV will currently fail with unresolved imports in the cloud
2. An application that doesn't use KV but where the compiler is not guaranteed to elide the dead imports will also fail with unresolved imports

and say what we propose to do about this (documentation, stubs in the cloud, etc.).

(I know this is a SIP and the details of cloud may be considered out of scope, but I reckon we should say _something._)

Reply:
This is at least partially addressed by https://github.com/fermyon/spin/pull/1048 (see the last paragraph in that document, in particular).  Perhaps it would be useful to reference that SIP from this one?

Reply:
At the moment, I'm afraid I don't feel I know enough from that paragraph about how we propose to handle this in the KV case, which is why I brought it up here.

The comment at the moment kind of feels like a side suggestion to an explanatory note about the rationale for a particular detail in a proposal that was primarily about improving error messages.  If it's a definite proposal, we could break it out into a top-level section about managing how changes propagate from Spin to Cloud, and make it more concrete (e.g. the timing of WIT iteration -> Cloud stubs -> Spin release).  That would certainly address my uncertainty.